### PR TITLE
Add support for parameters in paths

### PR DESCRIPTION
**Pull request by tonyarnold**

### CODE REVIEW

**Comment by swhitty** [comment]
Nice, this looks really flexible.

Do you have any idea on how you would expose the route & parameters to a handler?  These are the options I can see;

1. Handlers just reconstruct the route to extract parameters;
```swift
{ request in
   let route = HTTPRoute("GET /:id/hello/:food")
   let id = request.extractParameter(named: "id", within: route)
   return HTTPResponse(statusCode.ok)
}
```

2. Add the matched route to `HTTPRequest` before its passed to the handler
```swift
{ request in
   let id = request.route.pathParameters["id"]
   return HTTPResponse(statusCode.ok)
}
```

3. Update `protocol HTTPHandler` to include the matched route (larger breaking change):
```swift
public protocol HTTPHandler: Sendable {
    func handleRequest(_ request: HTTPRequest, for route: HTTPRoute) async throws -> HTTPResponse
}
```

4. Place the matched route in a `@TaskLocal` which handlers could inspect:
```swift
{ request in
   let id = HTTPRoute.current.pathParameters["id"]
   return HTTPResponse(statusCode.ok)
}
```

5. Use Parameter Packs (Swift 5.9+) to extract params based on the variadic arguments of the handler;
```swift
await server.appendRoute("GET /:id/hello/:food") { (id: Int, food: String) -> HTTPResponse in
   HTTPResponse(
      statusCode: .ok,
      body: "\(id * 2) \(food)".data(using: .utf8)!
   )
}
```

Eventually I would like to use a [macro](https://github.com/swhitty/FlyingFox/issues/89#issuecomment-2151916815 to match them to the named methods arguments, not ordered index.

I've had a [proved it out here](https://github.com/swhitty/FlyingFox/commit/2aba169584c1a924d566a4adb9304660cb98f3c2) and it should extend to query params, but I appreciate that once extended to headers and post params within the the body the index of the parameter loses its semantic meaning so another means may be required regardless.

**Comment by tonyarnold** [comment]
Honestly for now I think Option #2 is the simplest (dictionary subscript access). I don't think it needs to be any more complex than that.

> Eventually I would like to use a https://github.com/swhitty/FlyingFox/issues/89#issuecomment-2151916815 to match them to the named methods arguments, not ordered index.

That sounds great, but macros are nigh unusable at this point, given the increase in compile times that they cause.

**Comment by tonyarnold** [comment]
Okay, I've not added tests for this, but the TaskLocal ended up being far simpler than other approaches, so I've used that. 

There is now a method on `HTTPRequest`: 

```swift
func pathParameter(for identifier: String) -> String?
```

This can be used to look up the value of a URL parameter.

File: FlyingFox/Sources/HTTPRequest.swift [hunk 1]
```diff
@@ -88,4 +90,20 @@ extension HTTPRequest {
     var shouldKeepAlive: Bool {
         headers[.connection]?.caseInsensitiveCompare("keep-alive") == .orderedSame
     }
+
+    func pathParameter(for identifier: String) -> String? {
```
Comment (swhitty) [inline FlyingFox/Sources/HTTPRequest.swift:94]:
```suggestion
    public func pathParameter(for identifier: String) -> String? {
```

**Review by swhitty — APPROVED** [review]
not great,  I can make this method public and add some tests 🙂

Thank you

**Pull request merged**